src/paginationCalls/pageDescriptor.ts

```diff
@@ -103,6 +116,108 @@ export class PageDescriptor implements Descriptor {
     return stream;
   }
 
+  /**
+   * Create an async iterable which can be recursively called for data on-demand.
+   */
+  asyncIterate(
+    apiCall: GaxCall,
+    request: RequestType,
+    options: CallSettings
+  ): AsyncIterable<{} | undefined> {
+    const iterable = this.createIterator(options);
+    const funcPromise =
+      typeof apiCall === 'function' ? Promise.resolve(apiCall) : apiCall;
+    funcPromise
+      .then((func: GaxCall) => {
+        this.resolveParams(request, func, options);
+      })
+      .catch(error => {
+        throw new Error(error);
+      });
+    return iterable;
+  }
+
+  createIterator(options: CallSettings): AsyncIterable<{} | undefined> {
+    const requestPageTokenFieldName = this.requestPageTokenField;
+    const responsePageTokenFieldName = this.responsePageTokenField;
+    const resourceField = this.resourceField;
+    const self = this;
+    const getNextPageRequest = this.getNextPageRequest;
+    const asyncIterable = {
+      [Symbol.asyncIterator]() {
+        const funcPromise = new Promise((resolve, reject) => {
+          self.resolveFunction = resolve;
+        });
+        const requestPromise = new Promise((resolve, reject) => {
+          self.resolveRequest = resolve;
+        });
+        const cache: Array<{}> = [];
+        let nextPageRequest: RequestType | null = {};
+        let firstCall = true;
+        return {
+          async next() {
+            const ongoingCall = new call.OngoingCallPromise(options.promise);
+            const func = (await funcPromise) as SimpleCallbackFunction;
+            const request = (await requestPromise) as RequestType;
+            if (cache.length > 0) {
+              const value = cache.shift();
+              return Promise.resolve({done: false, value});
+            }
+            if (!firstCall && !nextPageRequest) {
+              return Promise.resolve({done: true, value: -1});
+            }
+            nextPageRequest = await getNextPageRequest(
+              cache,
+              func,
+              firstCall ? request : nextPageRequest!,
+              ongoingCall,
+              requestPageTokenFieldName,
+              responsePageTokenFieldName,
+              resourceField
+            );
+            firstCall = false;
+            return Promise.resolve({done: false, value: cache.shift()});
+          },
+        };
+      },
+    };
+    return asyncIterable; // return iterable
+  }
+
+  async getNextPageRequest(
+    cache: Array<{}>,
```

Comment:
7 parameters to the function is too much, we need to find a way to simplify the code here.

Reply:
It makes sense to store everything we need as fields of `PageDescriptor` (including `cache`), then we won't need to pass anything to the methods.

Reply:
sure, that's a good idea, will move all needed params as descriptor fields